In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pickle

In [3]:
# Define the root directory where your CSV files are located
root_directory = 'C:/Users/barsi/OneDrive - UNSW/Desktop/Archit/A3ADVMLA/Data/itineraries_csv'

# Initialize an empty list to store DataFrames
df = []

# Walk through the directory and its subdirectories
for foldername, subfolders, filenames in os.walk(root_directory):
    for filename in filenames:
        if filename.endswith('.csv'):
            file_path = os.path.join(foldername, filename)
            df_temp = pd.read_csv(file_path)
            df.append(df_temp)

# Concatenate all the DataFrames into a single DataFrame
df = pd.concat(df, ignore_index=True)

print(df)


                                     legId  searchDate  flightDate  \
0         9ca0e81111c683bec1012473feefd28f  2022-04-16  2022-04-17   
1         98685953630e772a098941b71906592b  2022-04-16  2022-04-17   
2         98d90cbc32bfbb05c2fc32897c7c1087  2022-04-16  2022-04-17   
3         969a269d38eae583f455486fa90877b4  2022-04-16  2022-04-17   
4         980370cf27c89b40d2833a1d5afc9751  2022-04-16  2022-04-17   
...                                    ...         ...         ...   
13519994  e381378bc47d12a1a46d094b427b7429  2022-05-19  2022-06-03   
13519995  1a41964e03957c790540dda613a4e445  2022-05-19  2022-06-03   
13519996  c3ce778a21b9f474abdb0e0678f44164  2022-05-19  2022-06-03   
13519997  fb53d403fe371735d45859e6df7c7590  2022-05-19  2022-06-03   
13519998  88c7878b462d04872939d16b10d57d7d  2022-05-19  2022-06-03   

         startingAirport destinationAirport travelDuration  isBasicEconomy  \
0                    ATL                BOS        PT2H29M           False   
1  

In [4]:
df.shape

(13519999, 23)

In [5]:
pd.set_option('display.max_columns', None)

In [6]:
df.head(20)

legId  searchDate  flightDate startingAirport  \
0   9ca0e81111c683bec1012473feefd28f  2022-04-16  2022-04-17             ATL   
1   98685953630e772a098941b71906592b  2022-04-16  2022-04-17             ATL   
2   98d90cbc32bfbb05c2fc32897c7c1087  2022-04-16  2022-04-17             ATL   
3   969a269d38eae583f455486fa90877b4  2022-04-16  2022-04-17             ATL   
4   980370cf27c89b40d2833a1d5afc9751  2022-04-16  2022-04-17             ATL   
5   79eda9f841e226a1e2121d74211e595c  2022-04-16  2022-04-17             ATL   
6   9335fae376c38bb61263281779f469ec  2022-04-16  2022-04-17             ATL   
7   3904bf87f2d1daf334f1ae7e3b876028  2022-04-16  2022-04-17             ATL   
8   d93988734c44a3c075d9efe373352507  2022-04-16  2022-04-17             ATL   
9   562e7d5dd6ecbf1509c0c19711dbdca9  2022-04-16  2022-04-17             ATL   
10  c38a6e4b807d15541e5866676febcbec  2022-04-16  2022-04-17             ATL   
11  f66d72ba3a526576608b8402b5720726  2022-04-16  2022-04-17             ATL   
12  e7c4054e85cca9bc7134b3cc93deaf7c  2022-04-16  2022-04-17             ATL   
13  5fa8c0f8b25eb24bf2ddc39a498bcc37  2022-04-16  2022-04-17             ATL   
14  948d26b3e5658762cd62f842c87ae46e  2022-04-16  2022-04-17             ATL   
15  cb345693479e84838dfaac346b8e8d6a  2022-04-16  2022-04-17             ATL   
16  e1b95e4e6c997517f64ba5ea712bda22  2022-04-16  2022-04-17             ATL   
17  eaf033a044596f0a76b132a490927524  2022-04-16  2022-04-17             ATL   
18  721d9a2f66fe479e7c17b13e7ae0bb15  2022-04-16  2022-04-17             ATL   
19  a9f012defb9227f69bf76617b47f4c51  2022-04-16  2022-04-17             ATL   

   destinationAirport travelDuration  isBasicEconomy  isRefundable  isNonStop  \
0                 BOS        PT2H29M           False         False       True   
1                 BOS        PT2H30M           False         False       True   
2                 BOS        PT2H30M           False         False       True   
3                 BOS        PT2H32M           False         False       True   
4                 BOS        PT2H34M           False         False       True   
5                 BOS        PT2H38M           False         False       True   
6                 BOS        PT4H12M           False         False      False   
7                 BOS        PT5H18M           False         False      False   
8                 BOS        PT5H32M           False         False      False   
9                 BOS        PT6H38M           False         False      False   
10                BOS        PT4H46M           False         False      False   
11                BOS        PT5H45M           False         False      False   
12                BOS        PT5H59M           False         False      False   
13                BOS        PT7H18M           False         False      False   
14                BOS        PT8H10M           False         False      False   
15                BOS        PT5H39M           False         False      False   
16                BOS        PT2H38M           False         False       True   
17                BOS        PT4H17M           False         False      False   
18                BOS        PT4H36M           False         False      False   
19                BOS        PT4H45M           False         False      False   

    totalFare  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
0      248.60                947.0                        1650214620   
1      248.60                947.0                        1650191400   
2      248.60                947.0                        1650209700   
3      248.60                947.0                        1650218340   
4      248.60                947.0                        1650203940   
5      248.60                947.0                        1650206700   
6      251.10                956.0            1650198000||1650205620   
7      251.10                956.0            1650198000||1650208800   


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13519999 entries, 0 to 13519998
Data columns (total 23 columns):
 #   Column                             Dtype  
---  ------                             -----  
 0   legId                              object 
 1   searchDate                         object 
 2   flightDate                         object 
 3   startingAirport                    object 
 4   destinationAirport                 object 
 5   travelDuration                     object 
 6   isBasicEconomy                     bool   
 7   isRefundable                       bool   
 8   isNonStop                          bool   
 9   totalFare                          float64
 10  totalTravelDistance                float64
 11  segmentsDepartureTimeEpochSeconds  object 
 12  segmentsDepartureTimeRaw           object 
 13  segmentsArrivalTimeEpochSeconds    object 
 14  segmentsArrivalTimeRaw             object 
 15  segmentsArrivalAirportCode         object 
 16  segmentsDepartur

In [7]:
df.describe(include='all')

legId  searchDate  flightDate  \
count                           13519999    13519999    13519999   
unique                           1721518          32          92   
top     940f2444dc7683325b6237c50ec56d69  2022-05-02  2022-05-24   
freq                                  35      540011      285643   
mean                                 NaN         NaN         NaN   
std                                  NaN         NaN         NaN   
min                                  NaN         NaN         NaN   
25%                                  NaN         NaN         NaN   
50%                                  NaN         NaN         NaN   
75%                                  NaN         NaN         NaN   
max                                  NaN         NaN         NaN   

       startingAirport destinationAirport travelDuration isBasicEconomy  \
count         13519999           13519999       13519999       13519999   
unique              16                 16           1836              2   
top                LAX                LAX        PT1H25M          False   
freq           1352275            1295126          53758       12768661   
mean               NaN                NaN            NaN            NaN   
std                NaN                NaN            NaN            NaN   
min                NaN                NaN            NaN            NaN   
25%                NaN                NaN            NaN            NaN   
50%                NaN                NaN            NaN            NaN   
75%                NaN                NaN            NaN            NaN   
max                NaN                NaN            NaN            NaN   

       isRefundable isNonStop     totalFare  totalTravelDistance  \
count      13519999  13519999  1.352000e+07         1.256038e+07   
unique            2         2           NaN                  NaN   
top           False     False           NaN                  NaN   
freq       13519808   9564064           NaN                  NaN   
mean            NaN       NaN  3.737535e+02         1.569619e+03   
std             NaN       NaN  2.075496e+02         8.414888e+02   
min             NaN       NaN  2.397000e+01         9.700000e+01   
25%             NaN       NaN  2.235800e+02         8.620000e+02   
50%             NaN       NaN  3.446000e+02         1.392000e+03   
75%             NaN       NaN  4.876000e+02         2.376000e+03   
max             NaN       NaN  8.260610e+03         4.430000e+03   

       segmentsDepartureTimeEpochSeconds       segmentsDepartureTimeRaw  \
count                           13519999                       13519999   
unique                           1273965                        1374911   
top                           1653040800  2022-05-18T06:00:00.000-04:00   
freq                                1703                           1599   
mean                                 NaN                            NaN   
std                                  NaN                            NaN   
min                                  NaN                            NaN   
25%                                  NaN                            NaN   
50%                                  NaN                            NaN   
75%                                  NaN                            NaN   
max                                  NaN                            NaN   

       segmentsArrivalTimeEpochSeconds         segmentsArrivalTimeRaw  \
count                         13519999                       13519999   
unique                         1447133                        1513276   
top                         1653278340  2022-05-20T23:59:00.000-04:00   
freq                               508                            378   
mean                               NaN                            NaN   
std                                NaN                            NaN   
min                                NaN                            NaN   
25%            

Lets check if there are any duplicates present in the dataset

In [8]:
df.duplicated().sum()

0

There are zero dupliactes present in the dataset

Checking whether there are any null values present in the dataset:

In [9]:
df.isna().sum()

legId                                     0
searchDate                                0
flightDate                                0
startingAirport                           0
destinationAirport                        0
travelDuration                            0
isBasicEconomy                            0
isRefundable                              0
isNonStop                                 0
totalFare                                 0
totalTravelDistance                  959619
segmentsDepartureTimeEpochSeconds         0
segmentsDepartureTimeRaw                  0
segmentsArrivalTimeEpochSeconds           0
segmentsArrivalTimeRaw                    0
segmentsArrivalAirportCode                0
segmentsDepartureAirportCode              0
segmentsAirlineName                       0
segmentsAirlineCode                       0
segmentsEquipmentDescription         262676
segmentsDurationInSeconds                 0
segmentsDistance                     126985
segmentsCabinCode               

Nulls are present in 'segmentsDistance', 'segmentsEquipmentDescription' and 'totalTravelDistance'

Lets check for outliers in the target variable:

In [ ]:
sns.boxplot(x=df["totalFare"])

Outliers could be mostly those fares that are more than $4000. Let us check the number of trips that have such high fare.

In [10]:
len(df[df['totalFare']>4000])

75

There are only 75 such cases in the dataset. These could be genuine cases where the tickets are of first class or those that are searched very close to the flight date.

In [11]:
df[df['totalFare']>4000]['segmentsCabinCode'].str.contains('first').sum()

68

This confirms our hypothesis and hence, we won't discard these rows

Now lets filter the column which are more appropriate and which we are going to take for model building

In [121]:
cols=['legId', 'flightDate', 'startingAirport', 'destinationAirport', 'isRefundable', 'segmentsCabinCode','segmentsDepartureTimeRaw', 'totalFare']
data=df[cols]

In [122]:
data.head()

legId  flightDate startingAirport  \
0  9ca0e81111c683bec1012473feefd28f  2022-04-17             ATL   
1  98685953630e772a098941b71906592b  2022-04-17             ATL   
2  98d90cbc32bfbb05c2fc32897c7c1087  2022-04-17             ATL   
3  969a269d38eae583f455486fa90877b4  2022-04-17             ATL   
4  980370cf27c89b40d2833a1d5afc9751  2022-04-17             ATL   

  destinationAirport  isRefundable segmentsCabinCode  \
0                BOS         False             coach   
1                BOS         False             coach   
2                BOS         False             coach   
3                BOS         False             coach   
4                BOS         False             coach   

        segmentsDepartureTimeRaw  totalFare  
0  2022-04-17T12:57:00.000-04:00      248.6  
1  2022-04-17T06:30:00.000-04:00      248.6  
2  2022-04-17T11:35:00.000-04:00      248.6  
3  2022-04-17T13:59:00.000-04:00      248.6  
4  2022-04-17T09:59:00.000-04:00      248.6

In [123]:
data.shape

(13519999, 8)

Lets look at the categories of 'segmentsCabinCode'

In [124]:
data.segmentsCabinCode.value_counts()

segmentsCabinCode
coach||coach                                   8528140
coach                                          3940119
coach||coach||coach                             950943
coach||coach||coach||coach                       37426
premium coach                                     9976
premium coach||coach                              9486
first||coach                                      9030
coach||premium coach                              8097
first||first                                      6286
coach||first                                      4718
first                                             3748
business                                          2092
premium coach||premium coach                      2014
first||first||first                               1247
premium coach||coach||coach                       1244
coach||coach||premium coach                        915
business||business                                 669
first||coach||coach                            

In [125]:
# Define a custom function to extract the common part
def extract_common(value):
    parts = value.split('||')
    common_part = parts[0]
    for part in parts:
        if part != common_part:
            return value  # Return the original value if any part is different
    return common_part

In [126]:
data['segmentsCabinCode'] = data['segmentsCabinCode'].apply(extract_common)

C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\3860362450.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['segmentsCabinCode'] = data['segmentsCabinCode'].apply(extract_common)


In [127]:
col = ['segmentsCabinCode', 'startingAirport', 'destinationAirport']
for col in col:
    print(data[col].value_counts())

segmentsCabinCode
coach                                         13456628
premium coach                                    11991
first                                            11281
premium coach||coach                              9486
first||coach                                      9030
coach||premium coach                              8097
coach||first                                      4718
business                                          2761
premium coach||coach||coach                       1244
coach||coach||premium coach                        915
first||coach||coach                                427
coach||premium coach||coach                        392
business||coach                                    377
coach||first||first                                350
coach||business                                    335
coach||coach||first                                330
first||first||coach                                300
coach||first||coach                            

We will encode all these categorical values into numeric later to feed to our machine learning model.

In [128]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13519999 entries, 0 to 13519998
Data columns (total 8 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   legId                     object 
 1   flightDate                object 
 2   startingAirport           object 
 3   destinationAirport        object 
 4   isRefundable              bool   
 5   segmentsCabinCode         object 
 6   segmentsDepartureTimeRaw  object 
 7   totalFare                 float64
dtypes: bool(1), float64(1), object(6)
memory usage: 734.9+ MB


In [129]:
data.head(20)

legId  flightDate startingAirport  \
0   9ca0e81111c683bec1012473feefd28f  2022-04-17             ATL   
1   98685953630e772a098941b71906592b  2022-04-17             ATL   
2   98d90cbc32bfbb05c2fc32897c7c1087  2022-04-17             ATL   
3   969a269d38eae583f455486fa90877b4  2022-04-17             ATL   
4   980370cf27c89b40d2833a1d5afc9751  2022-04-17             ATL   
5   79eda9f841e226a1e2121d74211e595c  2022-04-17             ATL   
6   9335fae376c38bb61263281779f469ec  2022-04-17             ATL   
7   3904bf87f2d1daf334f1ae7e3b876028  2022-04-17             ATL   
8   d93988734c44a3c075d9efe373352507  2022-04-17             ATL   
9   562e7d5dd6ecbf1509c0c19711dbdca9  2022-04-17             ATL   
10  c38a6e4b807d15541e5866676febcbec  2022-04-17             ATL   
11  f66d72ba3a526576608b8402b5720726  2022-04-17             ATL   
12  e7c4054e85cca9bc7134b3cc93deaf7c  2022-04-17             ATL   
13  5fa8c0f8b25eb24bf2ddc39a498bcc37  2022-04-17             ATL   
14  948d26b3e5658762cd62f842c87ae46e  2022-04-17             ATL   
15  cb345693479e84838dfaac346b8e8d6a  2022-04-17             ATL   
16  e1b95e4e6c997517f64ba5ea712bda22  2022-04-17             ATL   
17  eaf033a044596f0a76b132a490927524  2022-04-17             ATL   
18  721d9a2f66fe479e7c17b13e7ae0bb15  2022-04-17             ATL   
19  a9f012defb9227f69bf76617b47f4c51  2022-04-17             ATL   

   destinationAirport  isRefundable segmentsCabinCode  \
0                 BOS         False             coach   
1                 BOS         False             coach   
2                 BOS         False             coach   
3                 BOS         False             coach   
4                 BOS         False             coach   
5                 BOS         False             coach   
6                 BOS         False             coach   
7                 BOS         False             coach   
8                 BOS         False             coach   
9                 BOS         False             coach   
10                BOS         False             coach   
11                BOS         False             coach   
12                BOS         False             coach   
13                BOS         False             coach   
14                BOS         False             coach   
15                BOS         False             coach   
16                BOS         False             coach   
17                BOS         False             coach   
18                BOS         False             coach   
19                BOS         False             coach   

                             segmentsDepartureTimeRaw  totalFare  
0                       2022-04-17T12:57:00.000-04:00     248.60  
1                       2022-04-17T06:30:00.000-04:00     248.60  
2                       2022-04-17T11:35:00.000-04:00     248.60  
3                       2022-04-17T13:59:00.000-04:00     248.60  
4                       2022-04-17T09:59:00.000-04:00     248.60  
5                       2022-04-17T10:45:00.000-04:00     248.60  
6   2022-04-17T08:20:00.000-04:00||2022-04-17T10:2...     251.10  
7   2022-04-17T08:20:00.000-04:00||2022-04-17T11:2...     251.10  
8   2022-04-17T07:00:00.000-04:00||2022-04-17T10:2...     251.10  
9   2022-04-17T07:00:00.000-04:00||2022-04-17T11:2...     251.10  
10  2022-04-17T05:30:00.000-04:00||2022-04-17T08:3...     252.60  
11  2022-04-17T14:39:00.000-04:00||2022-04-17T17:0...     252.60  
12  2022-04-17T18:00:00.000-04:00||2022-04-17T20:3...     252.60  
13  2022-04-17T14:39:00.000-04:00||2022-04-17T18:4...     252.60  
14  2022-04-17T06:10:00.000-04:00||2022-04-17T10:5...     252.60  
15  2022-04-17T05:26:00.000-04:00||2022-04-17T08:0...     290.58  
16                      2022-04-17T10:45:00.000-04:00     300.10  
17  2022-04-17T18:15:00.000-04:00||2022-04-17T20:2...     302.11  
18  2022-04-17T11:02:00.000-04:00||2022-04-17T13:2...     302.11  
19  2022-04-17T16:18:00.000-04:00||2022-04-17T18:5...     302.11

Feature engineering and encoding:

In [130]:
#converting flighDate into datetime datatype:
data['flightDate'] = pd.to_datetime(data['flightDate'])

#Extracting and creating new features out of the date variable
data['flightYear'] = data['flightDate'].dt.year
data['flightMonth'] = data['flightDate'].dt.month
data['flightDay'] = data['flightDate'].dt.day

C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\2848235183.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['flightDate'] = pd.to_datetime(data['flightDate'])
C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\2848235183.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['flightYear'] = data['flightDate'].dt.year
C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\2848235183.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_inde

In [131]:
data.tail()

legId flightDate startingAirport  \
13519994  e381378bc47d12a1a46d094b427b7429 2022-06-03             SFO   
13519995  1a41964e03957c790540dda613a4e445 2022-06-03             SFO   
13519996  c3ce778a21b9f474abdb0e0678f44164 2022-06-03             SFO   
13519997  fb53d403fe371735d45859e6df7c7590 2022-06-03             SFO   
13519998  88c7878b462d04872939d16b10d57d7d 2022-06-03             SFO   

         destinationAirport  isRefundable segmentsCabinCode  \
13519994                PHL         False             coach   
13519995                PHL         False             coach   
13519996                PHL         False             coach   
13519997                PHL         False             coach   
13519998                PHL         False             coach   

                                   segmentsDepartureTimeRaw  totalFare  \
13519994  2022-06-03T06:00:00.000-07:00||2022-06-03T13:0...      836.6   
13519995  2022-06-03T05:10:00.000-07:00||2022-06-03T09:5...      837.6   
13519996  2022-06-03T13:06:00.000-07:00||2022-06-03T20:5...      837.6   
13519997  2022-06-03T07:30:00.000-07:00||2022-06-03T11:1...      837.6   
13519998                      2022-06-03T12:59:00.000-07:00     1649.6   

          flightYear  flightMonth  flightDay  
13519994        2022            6          3  
13519995        2022            6          3  
13519996        2022            6          3  
13519997        2022            6          3  
13519998        2022            6          3

lets get departure time from the segmentsDepartureTimeRaw feature:

In [132]:
# Function to extract the first time from the time_values column
def extract_time(time_value):
    times = time_value.split("||")
    if times:
        return times[0].split("T")[1]
    return ""

# Apply the function to the DataFrame and create a new column with the first times
data['departureTime'] = data['segmentsDepartureTimeRaw'].apply(extract_time)

C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\1200456476.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['departureTime'] = data['segmentsDepartureTimeRaw'].apply(extract_time)


In [133]:
data.head()

legId flightDate startingAirport  \
0  9ca0e81111c683bec1012473feefd28f 2022-04-17             ATL   
1  98685953630e772a098941b71906592b 2022-04-17             ATL   
2  98d90cbc32bfbb05c2fc32897c7c1087 2022-04-17             ATL   
3  969a269d38eae583f455486fa90877b4 2022-04-17             ATL   
4  980370cf27c89b40d2833a1d5afc9751 2022-04-17             ATL   

  destinationAirport  isRefundable segmentsCabinCode  \
0                BOS         False             coach   
1                BOS         False             coach   
2                BOS         False             coach   
3                BOS         False             coach   
4                BOS         False             coach   

        segmentsDepartureTimeRaw  totalFare  flightYear  flightMonth  \
0  2022-04-17T12:57:00.000-04:00      248.6        2022            4   
1  2022-04-17T06:30:00.000-04:00      248.6        2022            4   
2  2022-04-17T11:35:00.000-04:00      248.6        2022            4   
3  2022-04-17T13:59:00.000-04:00      248.6        2022            4   
4  2022-04-17T09:59:00.000-04:00      248.6        2022            4   

   flightDay       departureTime  
0         17  12:57:00.000-04:00  
1         17  06:30:00.000-04:00  
2         17  11:35:00.000-04:00  
3         17  13:59:00.000-04:00  
4         17  09:59:00.000-04:00

In [134]:
# Function to extract hours and minutes
def extract_hours(departure_time):
    parts = departure_time.split(':')
    if len(parts) >= 2:
        return int(parts[0])
    return None

def extract_minutes(departure_time):
    parts = departure_time.split(':')
    if len(parts) >= 3:
        minutes = parts[1].split('.')[0] 
        return int(minutes)
    return None

data['hour'] = data['departureTime'].apply(extract_hours)
data['minute'] = data['departureTime'].apply(extract_minutes)

C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\3848365346.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['hour'] = data['departureTime'].apply(extract_hours)
C:\Users\barsi\AppData\Local\Temp\ipykernel_16200\3848365346.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['minute'] = data['departureTime'].apply(extract_minutes)


In [135]:
data.head()

legId flightDate startingAirport  \
0  9ca0e81111c683bec1012473feefd28f 2022-04-17             ATL   
1  98685953630e772a098941b71906592b 2022-04-17             ATL   
2  98d90cbc32bfbb05c2fc32897c7c1087 2022-04-17             ATL   
3  969a269d38eae583f455486fa90877b4 2022-04-17             ATL   
4  980370cf27c89b40d2833a1d5afc9751 2022-04-17             ATL   

  destinationAirport  isRefundable segmentsCabinCode  \
0                BOS         False             coach   
1                BOS         False             coach   
2                BOS         False             coach   
3                BOS         False             coach   
4                BOS         False             coach   

        segmentsDepartureTimeRaw  totalFare  flightYear  flightMonth  \
0  2022-04-17T12:57:00.000-04:00      248.6        2022            4   
1  2022-04-17T06:30:00.000-04:00      248.6        2022            4   
2  2022-04-17T11:35:00.000-04:00      248.6        2022            4   
3  2022-04-17T13:59:00.000-04:00      248.6        2022            4   
4  2022-04-17T09:59:00.000-04:00      248.6        2022            4   

   flightDay       departureTime  hour  minute  
0         17  12:57:00.000-04:00    12      57  
1         17  06:30:00.000-04:00     6      30  
2         17  11:35:00.000-04:00    11      35  
3         17  13:59:00.000-04:00    13      59  
4         17  09:59:00.000-04:00     9      59

In [136]:
data = data.rename(columns={'hour': 'departureHour', 'minute': 'departureMinutes'})

In [137]:
data['isRefundable']=data['isRefundable'].astype(int)

In [138]:
data.head()

legId flightDate startingAirport  \
0  9ca0e81111c683bec1012473feefd28f 2022-04-17             ATL   
1  98685953630e772a098941b71906592b 2022-04-17             ATL   
2  98d90cbc32bfbb05c2fc32897c7c1087 2022-04-17             ATL   
3  969a269d38eae583f455486fa90877b4 2022-04-17             ATL   
4  980370cf27c89b40d2833a1d5afc9751 2022-04-17             ATL   

  destinationAirport  isRefundable segmentsCabinCode  \
0                BOS             0             coach   
1                BOS             0             coach   
2                BOS             0             coach   
3                BOS             0             coach   
4                BOS             0             coach   

        segmentsDepartureTimeRaw  totalFare  flightYear  flightMonth  \
0  2022-04-17T12:57:00.000-04:00      248.6        2022            4   
1  2022-04-17T06:30:00.000-04:00      248.6        2022            4   
2  2022-04-17T11:35:00.000-04:00      248.6        2022            4   
3  2022-04-17T13:59:00.000-04:00      248.6        2022            4   
4  2022-04-17T09:59:00.000-04:00      248.6        2022            4   

   flightDay       departureTime  departureHour  departureMinutes  
0         17  12:57:00.000-04:00             12                57  
1         17  06:30:00.000-04:00              6                30  
2         17  11:35:00.000-04:00             11                35  
3         17  13:59:00.000-04:00             13                59  
4         17  09:59:00.000-04:00              9                59

In [139]:
#Checking the count of refundable and non refundable planes:
data.isRefundable.value_counts()

isRefundable
0    13519808
1         191
Name: count, dtype: int64

It is surprising to see how few people have booked for non refundable flight ticket out of 13519999

In [140]:
data.tail()

legId flightDate startingAirport  \
13519994  e381378bc47d12a1a46d094b427b7429 2022-06-03             SFO   
13519995  1a41964e03957c790540dda613a4e445 2022-06-03             SFO   
13519996  c3ce778a21b9f474abdb0e0678f44164 2022-06-03             SFO   
13519997  fb53d403fe371735d45859e6df7c7590 2022-06-03             SFO   
13519998  88c7878b462d04872939d16b10d57d7d 2022-06-03             SFO   

         destinationAirport  isRefundable segmentsCabinCode  \
13519994                PHL             0             coach   
13519995                PHL             0             coach   
13519996                PHL             0             coach   
13519997                PHL             0             coach   
13519998                PHL             0             coach   

                                   segmentsDepartureTimeRaw  totalFare  \
13519994  2022-06-03T06:00:00.000-07:00||2022-06-03T13:0...      836.6   
13519995  2022-06-03T05:10:00.000-07:00||2022-06-03T09:5...      837.6   
13519996  2022-06-03T13:06:00.000-07:00||2022-06-03T20:5...      837.6   
13519997  2022-06-03T07:30:00.000-07:00||2022-06-03T11:1...      837.6   
13519998                      2022-06-03T12:59:00.000-07:00     1649.6   

          flightYear  flightMonth  flightDay       departureTime  \
13519994        2022            6          3  06:00:00.000-07:00   
13519995        2022            6          3  05:10:00.000-07:00   
13519996        2022            6          3  13:06:00.000-07:00   
13519997        2022            6          3  07:30:00.000-07:00   
13519998        2022            6          3  12:59:00.000-07:00   

          departureHour  departureMinutes  
13519994              6                 0  
13519995              5                10  
13519996             13                 6  
13519997              7                30  
13519998             12                59

In [141]:
#dropping unwanted features:
data.drop(['flightDate','legId','segmentsDepartureTimeRaw', 'departureTime'], axis=1, inplace = True)

In [142]:
data.head()

startingAirport destinationAirport  isRefundable segmentsCabinCode  \
0             ATL                BOS             0             coach   
1             ATL                BOS             0             coach   
2             ATL                BOS             0             coach   
3             ATL                BOS             0             coach   
4             ATL                BOS             0             coach   

   totalFare  flightYear  flightMonth  flightDay  departureHour  \
0      248.6        2022            4         17             12   
1      248.6        2022            4         17              6   
2      248.6        2022            4         17             11   
3      248.6        2022            4         17             13   
4      248.6        2022            4         17              9   

   departureMinutes  
0                57  
1                30  
2                35  
3                59  
4                59

In [143]:
data.columns

Index(['startingAirport', 'destinationAirport', 'isRefundable',
       'segmentsCabinCode', 'totalFare', 'flightYear', 'flightMonth',
       'flightDay', 'departureHour', 'departureMinutes'],
      dtype='object')

In [144]:
num_cols = ['isRefundable', 'flightYear', 'flightMonth', 'flightDay', 'departureHour', 'departureMinutes']
cat_cols = ['startingAirport', 'destinationAirport', 'segmentsCabinCode']

Now we are having all our data in the numeric format, the next essential step is to scale our data. Scaling the data will bring all the records on the same scale which will help model predict better.

Train Test Split:

In [145]:
X=data.drop('totalFare',axis=1)
y=data['totalFare']

In [146]:
X.columns

Index(['startingAirport', 'destinationAirport', 'isRefundable',
       'segmentsCabinCode', 'flightYear', 'flightMonth', 'flightDay',
       'departureHour', 'departureMinutes'],
      dtype='object')

In [147]:
y

0            248.6
1            248.6
2            248.6
3            248.6
4            248.6
             ...  
13519994     836.6
13519995     837.6
13519996     837.6
13519997     837.6
13519998    1649.6
Name: totalFare, Length: 13519999, dtype: float64

As we have a lot of data to build a mchine learning model, I have decided to split the data into traing and test in 7:3 ratio and further the traing data divided into training and validation in 4:1 ratio

In [148]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=8)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.20, random_state=8)

Baseline perfomance

In [149]:
y_mean=y_train.mean()
y_mean

373.72038191308945

In [150]:
y_base=np.full(y_train.shape,y_mean)
y_base

array([373.72038191, 373.72038191, 373.72038191, ..., 373.72038191,
       373.72038191, 373.72038191])

Import MSE and MAE from sklearn:

In [151]:
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae

In [152]:
print('Mean Squared Error: ', mse(y_train, y_base, squared=False))
print('Mean Absolute Error: ', mae(y_train, y_base))

Mean Squared Error:  207.48133694686462
Mean Absolute Error:  154.94370193105533


In [153]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


Pipeline for modelling:

In [154]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, TargetEncoder
from joblib import dump, load

In [155]:
#transforming numeric columns using num_transformer
num_transformer = Pipeline(
    steps=[
        ('scaler', StandardScaler())
    ]
)

In [156]:
#transforming categorical columns using cat_transformer
cat_transformer = Pipeline(
    steps=[
        ('target_encoder', TargetEncoder(random_state=8))
    ]
)

In [157]:
#preprocessor
preprocessor = ColumnTransformer(
    transformers=[
        ('num_transformer', num_transformer, num_cols),
        ('cat_transformer', cat_transformer, cat_cols),
    ]
)

In [165]:
preprocessor_pipe = Pipeline(
    steps=[
        ('preprocessor', preprocessor)
    ]
)
dump(preprocessor_pipe, 'preprocessor_pipe.joblib')

ValueError: Expected array-like (array or non-string sequence), got None

In [159]:
#implementing pipeline:
import xgboost as xgb
xgb_pipe = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('xgb', xgb.XGBRegressor(random_state=42))
    ]
)

In [160]:
#fitting the pipeline
xgb_pipe.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num_transformer',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  ['isRefundable', 'flightYear',
                                                   'flightMonth', 'flightDay',
                                                   'departureHour',
                                                   'departureMinutes']),
                                                 ('cat_transformer',
                                                  Pipeline(steps=[('target_encoder',
                                                                   TargetEncoder(random_state=8))]),
                                                  ['startingAirport',
                                                   'destinationAirport',
                                                   'se...
                              feature_types=None, gamma=None, grow_policy=None,
                              importance_type=None,
                              interaction_constraints=None, learning_rate=None,
                              max_bin=None, max_cat_threshold=None,
                              max_cat_to_onehot=None, max_delta_step=None,
                              max_depth=None, max_leaves=None,
                              min_child_weight=None, missing=nan,
                              monotone_constraints=None, multi_strategy=None,
                              n_estimators=None, n_jobs=None,
                              num_parallel_tree=None, random_state=42, ...))])

In [161]:
#predicting y_train
y_train_preds = xgb_pipe.predict(X_train)
print('Mean Squared Error: ' , mse(y_train, y_train_preds, squared=False))
print('Mean Absolute Error: ' ,mae(y_train, y_train_preds))

Mean Squared Error:  129.58641341468442
Mean Absolute Error:  89.89061581530777


In [162]:
#predicting y_val
y_val_preds = xgb_pipe.predict(X_val)
print('Mean Squared Error: ',mse(y_val, y_val_preds, squared=False))
print('Mean Absolute Error: ',mae(y_val, y_val_preds)) 

Mean Squared Error:  129.614159831286
Mean Absolute Error:  89.96549445743493


It can be observed that the xgb regressor performed well on the unseen validation set as it did not overfit on the unseen data and gave almost the same mse and mae scores. Now lets check for the test data

Checking on the test data: 

In [163]:
#predicting y_test
y_test_preds = xgb_pipe.predict(X_test)
print('Mean Squared Error: ',mse(y_test, y_test_preds, squared=False))
print('Mean Absolute Error: ',mae(y_test, y_test_preds))  

Mean Squared Error:  129.89859373678777
Mean Absolute Error:  89.93234165223402


Our xgb regressor model performed great on the unseen testing data as well. No overfitting is observed

Lets save our created model, so that it can be reused again:

In [167]:
#saving the model:
from joblib import dump
dump(xgb_pipe, 'xgb_pipeline.joblib')

['xgb_pipeline.joblib']